### PR TITLE
Fix HKL cursor info for 4D MD workspaces

### LINKED ISSUE #34986: HKL cursor info table in sliceviewer not working for 4D MD workspaces
**Describe the bug**
Out of range error from HKL cursor info table in sliceviewer for 4D MD workspaces 
```
Traceback (most recent call last):
  File "C:\MantidNightlyInstall\bin\lib\site-packages\matplotlib\cbook\__init__.py", line 307, in process
    func(*args, **kwargs)
  File "C:\MantidNightlyInstall\bin\lib\site-packages\mantidqt\widgets\sliceviewer\views\cursor.py", line 66, in on_mouse_move
    self.on_cursor_at(xdata, ydata)
  File "C:\MantidNightlyInstall\bin\lib\site-packages\mantidqt\widgets\sliceviewer\presenters\imageinfowidget.py", line 84, in _on_cursor_at_axesimage
    extra_cols = self._presenter.get_extra_image_info_columns(xdata, ydata)
  File "C:\MantidNightlyInstall\bin\lib\site-packages\mantidqt\widgets\sliceviewer\presenters\presenter.py", line 547, in get_extra_image_info_columns
    hkl = self.model.get_hkl_from_xyz(xdim, ydim, zdim, xdata, ydata, slice_point[zdim])
  File "C:\MantidNightlyInstall\bin\lib\site-packages\mantidqt\widgets\sliceviewer\models\model.py", line 582, in get_hkl_from_xyz
    hkl = basis_transform[:, xdim] * xdata + basis_transform[:, ydim] * ydata + basis_transform[:, zdim] * zdata
IndexError: index 3 is out of bounds for axis 1 with size 3
Traceback (most recent call last):
  File "C:\MantidNightlyInstall\bin\lib\site-packages\matplotlib\cbook\__init__.py", line 307, in process
    func(*args, **kwargs)
  File "C:\MantidNightlyInstall\bin\lib\site-packages\mantidqt\widgets\sliceviewer\views\cursor.py", line 66, in on_mouse_move
    self.on_cursor_at(xdata, ydata)
  File "C:\MantidNightlyInstall\bin\lib\site-packages\mantidqt\widgets\sliceviewer\presenters\imageinfowidget.py", line 84, in _on_cursor_at_axesimage
    extra_cols = self._presenter.get_extra_image_info_columns(xdata, ydata)
  File "C:\MantidNightlyInstall\bin\lib\site-packages\mantidqt\widgets\sliceviewer\presenters\presenter.py", line 547, in get_extra_image_info_columns
    hkl = self.model.get_hkl_from_xyz(xdim, ydim, zdim, xdata, ydata, slice_point[zdim])
  File "C:\MantidNightlyInstall\bin\lib\site-packages\mantidqt\widgets\sliceviewer\models\model.py", line 582, in get_hkl_from_xyz
    hkl = basis_transform[:, xdim] * xdata + basis_transform[:, ydim] * ydata + basis_transform[:, zdim] * zdata
IndexError: index 3 is out of bounds for axis 1 with size 3
Traceback (most recent call last):
  File "C:\MantidNightlyInstall\bin\lib\site-packages\matplotlib\cbook\__init__.py", line 307, in process
    func(*args, **kwargs)
  File "C:\MantidNightlyInstall\bin\lib\site-packages\mantidqt\widgets\sliceviewer\views\cursor.py", line 66, in on_mouse_move
    self.on_cursor_at(xdata, ydata)
  File "C:\MantidNightlyInstall\bin\lib\site-packages\mantidqt\widgets\sliceviewer\presenters\imageinfowidget.py", line 84, in _on_cursor_at_axesimage
    extra_cols = self._presenter.get_extra_image_info_columns(xdata, ydata)
  File "C:\MantidNightlyInstall\bin\lib\site-packages\mantidqt\widgets\sliceviewer\presenters\presenter.py", line 547, in get_extra_image_info_columns
    hkl = self.model.get_hkl_from_xyz(xdim, ydim, zdim, xdata, ydata, slice_point[zdim])
  File "C:\MantidNightlyInstall\bin\lib\site-packages\mantidqt\widgets\sliceviewer\models\model.py", line 582, in get_hkl_from_xyz
    hkl = basis_transform[:, xdim] * xdata + basis_transform[:, ydim] * ydata + basis_transform[:, zdim] * zdata
IndexError: index 3 is out of bounds for axis 1 with size 3
```

**To Reproduce**

(1) Make workspace
```
ws_4D = CreateMDWorkspace(Dimensions=4, Extents=[-1, 1, -1, 1, -1, 1, -1, 1], Names="E,H,K,L",
                              Frames='General Frame,HKL,HKL,HKL', Units='meV,r.l.u.,r.l.u.,r.l.u.')
expt_info_4D = CreateSampleWorkspace()
ws_4D.addExperimentInfo(expt_info_4D)
SetUB(ws_4D, 1, 1, 2, 90, 90, 120)
```
(2) Move the cursor onto the colorfill - note it assumes the first three axes are HKL (actually EHK) 
(3) Change viewing axes to include the last axis (L)
(4) Move cursor into colorfill to get above error

**Expected behavior**
Would exclude non-momentum axes when calculating HKL

**Screenshots**

**Platform/Version (please complete the following information):**
 - OS: [e.g. Windows, RHEL 7, Ubuntu, macOS]
 - OS Version:
 - Mantid Version [e.g. 6.0.0]

**Additional context**

### PR DESCRIPTION
**Pull request by robertapplin**
**Description of work.**
This PR fixes a bug in the SliceViewer when viewing 4D data. The HKL values are now correctly printed in the cursor info widget even when another dimension exists. The dimensions can now be in any order and it should still work e.g. any of EHKL, HEKL, HKEL, HKLE.

**To test:**
Follow the instructions in the attached issue.
You can also try swapping the order of the dimensions and check it is still working.

Fixes #34986

*This does not require release notes* because **the bug was not present in the previous release**

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by RichardWaiteSTFC** [comment]
@MialLewis can you test it for the following workspace `ws_slice` which has axes that are a mix of h,k,l
```
ws = CreateMDWorkspace(Dimensions='3', Extents='-3,3,-3,3,-3,3',
                   Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.',
                   Frames='HKL,HKL,HKL',
                   SplitInto='2', SplitThreshold='10')
expt_info = CreateSampleWorkspace()
ws.addExperimentInfo(expt_info)
SetUB(ws, 1,1,2,90,90,120)
BinMD(InputWorkspace=ws, AxisAligned=False,
    BasisVector0='[00L],r.l.u.,0,0,1',
    BasisVector1='[HH0],r.l.u.,1,1,0',
    BasisVector2='[-HH0],r.l.u.,-1,1,0',
    OutputExtents='-4,4,-4,4,-0.25,0.25',
    OutputBins='101,101,1', OutputWorkspace='ws_slice', NormalizeBasisVectors=False)
```
And perhaps similar 4D workspaces with non-Q dimensions included? Cheers!
Alternatively I'm happy to test this PR - whatever you prefer!

**Comment by MialLewis** [comment]
> @MialLewis can you test it for the following workspace `ws_slice` which has axes that are a mix of h,k,l
> 
> ```
> ws = CreateMDWorkspace(Dimensions='3', Extents='-3,3,-3,3,-3,3',
>                    Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.',
>                    Frames='HKL,HKL,HKL',
>                    SplitInto='2', SplitThreshold='10')
> expt_info = CreateSampleWorkspace()
> ws.addExperimentInfo(expt_info)
> SetUB(ws, 1,1,2,90,90,120)
> BinMD(InputWorkspace=ws, AxisAligned=False,
>     BasisVector0='[00L],r.l.u.,0,0,1',
>     BasisVector1='[HH0],r.l.u.,1,1,0',
>     BasisVector2='[-HH0],r.l.u.,-1,1,0',
>     OutputExtents='-4,4,-4,4,-0.25,0.25',
>     OutputBins='101,101,1', OutputWorkspace='ws_slice', NormalizeBasisVectors=False)
> ```
> 
> And perhaps similar 4D workspaces with non-Q dimensions included? Cheers! Alternatively I'm happy to test this PR - whatever you prefer!

yep, I'm happy to do it, I've pretty much finished testing and about to approve

**Comment by gemmaguest** [comment]
@robertapplin Should this be going into `release-next` instead of `main`?

**Comment by robertapplin** [comment]
> @robertapplin Should this be going into `release-next` instead of `main`?

Yes, thank you!

**Comment by MialLewis** [comment]
As discussed, if you run the following script, `H` should be equal to `K`. `L` also appears to be tracking the third slice point.

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

ws_4D = CreateMDWorkspace(Dimensions=4, Extents=[-1, 1, -1, 1, -1, 1, -1, 1], Names="E,H,K,L",
                              Frames='General Frame,HKL,HKL,HKL', Units='meV,r.l.u.,r.l.u.,r.l.u.')
expt_info_4D = CreateSampleWorkspace()
ws_4D.addExperimentInfo(expt_info_4D)
SetUB(ws_4D, 1, 1, 2, 90, 90, 120)
BinMD(InputWorkspace=ws_4D, AxisAligned=False,
    BasisVector0='[00L],r.l.u.,0,0,0,1',
    BasisVector1='[HH0],r.l.u.,0,1,1,0',
    BasisVector2='[-HH0],r.l.u.,0,-1,1,0',
    BasisVector3='[E],r.l.u.,1,0,0,0',
    OutputExtents='-1,1,-4,4,-4,4,-0.25,0.25',
    OutputBins='100,100,100,100', OutputWorkspace='ws_slice', NormalizeBasisVectors=False)
```

File: qt/python/mantidqt/mantidqt/widgets/sliceviewer/models/model.py [hunk 1]
```diff
@@ -576,13 +596,27 @@ def _cut_names(self, cut: str):
 
         return xcut_name, ycut_name, help_msg
 
-    def get_hkl_from_xyz(self, xdim, ydim, zdim, xdata, ydata, zdata):
+    @staticmethod
+    def get_dim_indices(slice_point: List[float], qdims: List[int], transpose: bool) -> Tuple[int]:
+        """Returns the indices of the selected x and y axes. It also returns the index of the first non x or y axis."""
+        xdim, ydim = WorkspaceInfo.display_indices(slice_point, transpose)
+        # Get the index of the first axis which is not x or y. Here it is called z.
+        zdim = next(i for i, v in enumerate(slice_point) if v is not None and i in qdims)
```
Comment (RichardWaiteSTFC) [inline qt/python/mantidqt/mantidqt/widgets/sliceviewer/models/model.py:604]: Not sure this is working as expected - this assumes there is only one Q-dim with a slicepoint (see screenshot). 
`xdim` and `ydim` may be non-Q dimensions as well - but I think that's handled OK...

File: qt/python/mantidqt/mantidqt/widgets/sliceviewer/models/model.py [hunk 2]
```diff
@@ -576,13 +596,27 @@ def _cut_names(self, cut: str):
...
-        if basis_transform is not None:
-            hkl = basis_transform[:, xdim] * xdata + basis_transform[:, ydim] * ydata + basis_transform[:, zdim] * zdata
-        else:
-            hkl = {0.0, 0.0, 0.0}
-        return hkl
+        if basis_transform is None:
+            return 0.0, 0.0, 0.0
+
+        # Get the values for h, k and l
+        hkl_point = tuple(full_point[q_i] for q_i in qdims)
```
Comment (RichardWaiteSTFC) [inline qt/python/mantidqt/mantidqt/widgets/sliceviewer/models/model.py:614]: I think this is not the HKL but actually the point in the sliceviewer projection coordinates (see screenshot)

File: qt/python/mantidqt/mantidqt/widgets/sliceviewer/models/model.py [hunk 3]
```diff
@@ -576,13 +596,27 @@ def _cut_names(self, cut: str):
...
+
+        h_i, k_i, l_i = np.argsort([xdim, ydim, zdim])
```
Comment (RichardWaiteSTFC) [inline qt/python/mantidqt/mantidqt/widgets/sliceviewer/models/model.py:616]: Again, `h_i` etc. correspond to the index of the `proj_matrix` column correpsonding to the index of the elements in `hkl_point` - perhaps `icol_x` etc. would be better names?

**Review by RichardWaiteSTFC — CHANGES_REQUESTED** [review]
I think this is very close and actually works as expected for the workspaces tested because now you get the full point (HKL and any non_Q dimension positions) from a combination of the slicepoints and `xdim` & `ydim` however I think there is still a bug here.
In the screenshot (using the workspace @MialLewis posted) - the `zdim` passed to the `argsort` seems to be for a non-Q axis (it should be 2). Possibly because of a problem in `get_dim_indices` which @DannyHindson may have also talked to you about?
image

File: qt/python/mantidqt/mantidqt/widgets/sliceviewer/models/model.py [hunk 4]
```diff
@@ -535,6 +531,30 @@ def get_proj_matrix(self):
             return None
 
     # private api
+
+    @staticmethod
+    def _get_proj_matrix_from_basis_vectors(ws, proj_matrix):
+        ndims = ws.getNumDims()
+
+        # Array of True or False depending on if a dimension is in Q
+        qdims = np.array([ws.getDimension(i).getMDFrame().isQ() for i in range(ndims)])
```
Comment (RichardWaiteSTFC) [inline qt/python/mantidqt/mantidqt/widgets/sliceviewer/models/model.py:540]: Minor point, but as the `qdims` is something constant about the workspace we should probably store this as an attribute as we use it again when determining the HKL and maybe have some private methods to e.g. `get_index_nonQ_dims` etc. - what do you think?

**Comment by RichardWaiteSTFC** [comment]
Ah I think I see the problem - we no longer need the argsort (added to solve out-of range error) - the order of the elements in `hkl_point` (as discussed not actually HKL) are in the order of the workspace basis vectors (and columns in `proj_matrix`).

File: qt/python/mantidqt/mantidqt/widgets/sliceviewer/views/dimensionwidget.py [hunk 5]
```diff
@@ -96,6 +98,24 @@ def change_dims(self, number):
...
+    def n_dims(self):
+        """Returns the number of dimensions"""
+        return len(self.dims)
+
+    def q_dim_indices(self):
+        """Returns an array of indices for the Q dimensions"""
+        dims_i = np.arange(0, self.n_dims(), 1)
+
+        # Return the indices of dimensions which have Q units
+        return dims_i[self.qflags]
```
Comment (RichardWaiteSTFC) [inline qt/python/mantidqt/mantidqt/widgets/sliceviewer/views/dimensionwidget.py:110]: minor but could this be simplified to 
`return np.where(self.qflags)[0]`

**Comment by robertapplin** [comment]
@RichardWaiteSTFC I have created a separate issue to deal with the second bug https://github.com/mantidproject/mantid/issues/35095 . Hopefully dealing with them separately will make things a bit easier

File: qt/python/mantidqt/mantidqt/widgets/sliceviewer/models/model.py [hunk 6]
```diff
@@ -576,13 +576,16 @@ def _cut_names(self, cut: str):
...
-        if basis_transform is not None:
-            hkl = basis_transform[:, xdim] * xdata + basis_transform[:, ydim] * ydata + basis_transform[:, zdim] * zdata
-        else:
-            hkl = {0.0, 0.0, 0.0}
-        return hkl
+        if basis_transform is None:
+            return 0.0, 0.0, 0.0
+
+        # Get the values for the h, k and l projections
+        hkl_projection = tuple(full_point[q_i] for q_i in qdims_i)
```
Comment (RichardWaiteSTFC) [inline qt/python/mantidqt/mantidqt/widgets/sliceviewer/models/model.py:586]: Really minor but I think `hkl_projection` is not a good name, perhaps  `q_xyz`? Recall this is the Q vector in the frame of the sliceviewer axes.

File: qt/python/mantidqt/mantidqt/widgets/sliceviewer/models/model.py [hunk 7]
```diff
@@ -576,13 +576,16 @@ def _cut_names(self, cut: str):
 
         return xcut_name, ycut_name, help_msg
 
-    def get_hkl_from_xyz(self, xdim, ydim, zdim, xdata, ydata, zdata):
+    def get_hkl_from_full_point(self, full_point: List[float], qdims_i: List[int]):
```
Comment (RichardWaiteSTFC) [inline qt/python/mantidqt/mantidqt/widgets/sliceviewer/models/model.py:579]: As discussed offline, the `qdims_i` are actually a property that can be determined from the workspace and so I think ultimately belongs in the model (not retrieved from the data view dimension widget and passed via the presenter) - but I'm happy for that to be addressed in #35095

**Review by RichardWaiteSTFC — CHANGES_REQUESTED** [review]
Looks good, thanks for splitting up the issue - this is a good fix to get in for the release and a good addition to the tests. If you could perhaps just rename one variable then I'll approve (sorry very nit-picking of me)!

**Review by RichardWaiteSTFC — APPROVED** [review]
Great job!

**Pull request merged**